### PR TITLE
LogUnit server GC phase 1: refactoring, params and add skeleton codes

### PR DESCRIPTION
**Pull request by WenbinZhu**
## Overview

Pull request on the feature branch:

LogUnit server GC phase 1: refactoring, extract/add parameters and add add skeleton codes for compaction.

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java [hunk 1]
```diff
@@ -662,6 +664,22 @@ private EventLoopGroup getNewBossGroup() {
...
+     * @return an instance of {@link StreamLogParams}
+     */
+    public StreamLogParams getStreamLogParams() {
+        return StreamLogParams.builder()
+                .logPath(getServerConfig(String.class, "--log-path"))
+                .verifyChecksum(!getServerConfig(Boolean.class, "--no-verify"))
+                .build();
+    }
+
+    public StreamLogDataStore getStreamLogDataStore() {
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java]: Is singleton better here?

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/AbstractLogSegment.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    @NonNull
+    final FileChannel writeChannel;
+
+    @NonNull
+    final FileChannel readChannel;
+
+    // TODO: persist revision in header.
+    // Used for checking equality between stream log and garbage log segment in recovery.
+    long compactRevision;
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/AbstractLogSegment.java]: Need more detailed explanation.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+public interface CompactionPolicy {
+
+    /**
+     * Returns a list of segments selected for compaction based on
+     * the compaction policy implementation.
+     *
+     * @param compactibleSegments all unprotected segments that can be selected for compaction.
+     * @return a list of segments selected for compaction.
+     */
+    List<StreamLogSegment> segmentsToCompact(List<StreamLogSegment> compactibleSegments);
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java]: It is better to rename the function. "segmentsToCompact" sounds like a noun. Probably "compact" is an accurate name.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    static CompactionPolicy getPolicy(StreamLogParams params) {
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java]: Why use reflection? StreamLogParams could specify the policy in other ways, e.g. Enum.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/AbstractLogSegment.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    // public abstract LogData read(long address);
+
+    public abstract void append(long address, LogData entry);
+
+    // public Iterator<LogData> iterator() {
+    //     throw new NotImplementedException();
+    // }
+
+    public abstract void close();
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/AbstractLogSegment.java]: Why abstract function?

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+
+/**
+ * The interface of stream log compaction policy, which decides
+ * which log segments will be selected for compaction.
+ *
+ * Created by WenbinZhu on 5/22/19.
+ */
+public interface CompactionPolicy {
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java]: CompactionPolicy -> ICompactionPolicy

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentManager.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * Created by WenbinZhu on 5/28/19.
+ */
+@RequiredArgsConstructor
+public class SegmentManager {
+
+    private final StreamLogParams logParams;
+
+    @Getter
+    private final Map<Long, StreamLogSegment> streamLogSegments = new ConcurrentHashMap<>();
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentManager.java]: More detailed explanation on the two variables, including what the key means.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java (hunk 1 above)
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java]: Right now I don't see a strong reason to make it singleton.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/AbstractLogSegment.java (hunk 2 above)
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/AbstractLogSegment.java]: Will add when implementing it.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java (hunk 3 above)
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java]: Renamed, added a "get" prefix

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java (hunk 4 above)
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java]: First, the policy will be later specified by user on server startup, it's easier to use string. 
Second, even the StreamLogParams use Enum for policy, the implementation of the policy is not, so we have only two ways: `if...else` or reflection.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/AbstractLogSegment.java (hunk 5 above)
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/AbstractLogSegment.java]: StreamLogSegment and GrabageLogSegment may have different resources to close, so it's an abstract, if later implementation finishes and turns out they are the same, I will change it to non-abstract.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java (hunk 6 above)
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java]: The `I` prefix is an old-fashion pattern and we're avoiding it in recent patches. Slava had a strong argument on this.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentManager.java (hunk 7 above)
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentManager.java]: Yes, will be added once it's finalized.

**Pull request merged**